jetty-core/jetty-deploy/src/main/java/org/eclipse/jetty/deploy/providers/ContextProvider.java

```diff
@@ -589,56 +628,397 @@ else if (StringUtil.asciiStartsWithIgnoreCase(contextPath, "root-"))
         {
             int dash = contextPath.indexOf('-');
             String virtual = contextPath.substring(dash + 1);
-            context.setVirtualHosts(Arrays.asList(virtual.split(",")));
+            contextHandler.setVirtualHosts(Arrays.asList(virtual.split(",")));
             contextPath = "/";
         }
 
         // Ensure "/" is Prepended to all context paths.
         if (contextPath.charAt(0) != '/')
             contextPath = "/" + contextPath;
 
-        // Set the display name and context Path
-        context.setDisplayName(basename);
-        context.setContextPath(contextPath);
+        if (LOG.isDebugEnabled())
+            LOG.debug("ContextHandler {} initialized with displayName: {}", contextHandler, basename);
+        contextHandler.setDisplayName(basename);
+        if (LOG.isDebugEnabled())
+            LOG.debug("ContextHandler {} initialized with contextPath: {}", contextHandler, contextPath);
+        contextHandler.setContextPath(contextPath);
+    }
+
+    /**
+     * Apply the main deployable heuristics referenced in the main javadoc
+     * for this class.
+     *
+     * @param unit the unit of deployment to interrogate.
+     * @return the main deployable path
+     */
+    @Override
+    protected Path getMainDeploymentPath(Unit unit)
+    {
+        // XML always win.
+        List<Path> xmls = unit.getPaths().keySet().stream()
+            .filter(FileID::isXml)
```

Comment:
Need to filter out removed paths as well

Reply:
This uses the new `Unit.getLivePaths()`, which filters out REMOVED state paths.